src/types.jl

```diff
@@ -1,143 +1,219 @@
-  default_finaliser =
-    Model == :NeuralNetworkBinaryClassifier ? Flux.σ : Flux.softmax
-  default_loss =
-    Model == :NeuralNetworkBinaryClassifier ? Flux.binarycrossentropy : Flux.crossentropy
-
-  quote
-      mutable struct $Model{B,F,O,L} <: MLJFluxProbabilistic
-          builder::B
-          finaliser::F
-          optimiser::O   # mutable struct from Flux/src/optimise/optimisers.jl
-          loss::L        # can be called as in `loss(yhat, y)`
-          epochs::Int    # number of epochs
-          batch_size::Int  # size of a batch
-          lambda::Float64  # regularization strength
-          alpha::Float64   # regularizaton mix (0 for all l2, 1 for all l1)
-          rng::Union{AbstractRNG,Int64}
-          optimiser_changes_trigger_retraining::Bool
-          acceleration::AbstractResource  # eg, `CPU1()` or `CUDALibs()`
-      end
-
-      function $Model(
-          ;builder::B=$default_builder_ex,
-          finaliser::F=$default_finaliser,
-          optimiser::O=Optimisers.Adam(),
-          loss::L=$default_loss,
-          epochs=10,
-          batch_size=1,
-          lambda=0,
-          alpha=0,
-          rng=Random.default_rng(),
-          optimiser_changes_trigger_retraining=false,
-          acceleration=CPU1(),
-          ) where {B,F,O,L}
-
-          model = $Model{B,F,O,L}(
-              builder,
-              finaliser,
-              optimiser,
-              loss,
-              epochs,
-              batch_size,
-              lambda,
-              alpha,
-              rng,
-              optimiser_changes_trigger_retraining,
-              acceleration,
-          )
-
-          message = clean!(model)
-          isempty(message) || @warn message
-
-          return model
-      end
-
-  end |> eval
+const MLJFluxModel = Union{MLJFluxProbabilistic, MLJFluxDeterministic}
+
+for Model in [:NeuralNetworkClassifier, :NeuralNetworkBinaryClassifier]
+
+    # default settings that are not equal across models
+    default_finaliser =
+        Model == :NeuralNetworkBinaryClassifier ? Flux.σ : Flux.softmax
+    default_loss =
+        Model == :NeuralNetworkBinaryClassifier ? Flux.binarycrossentropy :
+        Flux.crossentropy
+
+    quote
+        mutable struct $Model{B, F, O, L} <: MLJFluxProbabilistic
+            builder::B
+            finaliser::F
+            optimiser::O   # mutable struct from Flux/src/optimise/optimisers.jl
+            loss::L        # can be called as in `loss(yhat, y)`
+            epochs::Int    # number of epochs
+            batch_size::Int  # size of a batch
+            lambda::Float64  # regularization strength
+            alpha::Float64   # regularizaton mix (0 for all l2, 1 for all l1)
+            rng::Union{AbstractRNG, Int64}
+            optimiser_changes_trigger_retraining::Bool
+            acceleration::AbstractResource  # eg, `CPU1()` or `CUDALibs()`
+            embedding_dims::Dict{Symbol, Real}
```

Comment:
I don't see a reason for an abstract type here:

```suggestion
            embedding_dims::Dict{Symbol, Int}
```

Reply:
@ablaom I handle differently depending on whether it's an integer or float as in the docs.